core/src/test/java/org/elasticsearch/cluster/action/shard/ShardStateActionTests.java

```diff
@@ -293,6 +291,41 @@ public void onFailure(Throwable t) {
         assertTrue(failure.get());
     }
 
+    public void testShardNotFound() throws InterruptedException {
+        final String index = "test";
+
+        clusterService.setState(stateWithStartedPrimary(index, true, randomInt(5)));
+
+        String indexUUID = clusterService.state().metaData().index(index).getIndexUUID();
+
+        AtomicBoolean success = new AtomicBoolean();
+        CountDownLatch latch = new CountDownLatch(1);
+
+        ShardRouting failedShard = getRandomShardRouting(index);
+        RoutingTable routingTable = RoutingTable.builder(clusterService.state().getRoutingTable()).remove(index).build();
+        clusterService.setState(ClusterState.builder(clusterService.state()).routingTable(routingTable));
```

Comment:
how about inlining this with multiple commands and check that everything works ok in batch mode? This can be a separate test...